packages/libs/error-stack/src/fmt/nightly.rs

```diff
@@ -0,0 +1,98 @@
+use alloc::{string::String, vec, vec::Vec};
+
+use crate::fmt::{Emit, Snippet};
+
+/// `nightly` experimental type, which is used during the formatting of [`Debug`] context via the
+/// [`Provider`] api.
+///
+/// # Example
+///
+/// ```
+/// use std::io::{Error, ErrorKind};
+///
+/// use error_stack::{fmt::DebugDiagnostic, report};
+///
+/// let report = report!(Error::from(ErrorKind::InvalidInput)) //
+///     .attach(DebugDiagnostic::next("Hello!"));
+///
+/// # owo_colors::set_override(true);
+/// # fn render(value: String) -> String {
+/// #     let backtrace = regex::Regex::new(r"Backtrace No\. (\d+)\n(?:  .*\n)*  .*").unwrap();
+/// #     let backtrace_info = regex::Regex::new(r"backtrace with (\d+) frames \((\d+)\)").unwrap();
+/// #
+/// #     let value = backtrace.replace_all(&value, "Backtrace No. $1\n  [redacted]");
+/// #     let value = backtrace_info.replace_all(value.as_ref(), "backtrace with [n] frames ($2)");
+/// #
+/// #     ansi_to_html::convert_escaped(value.as_ref()).unwrap()
+/// # }
+/// #
+/// # expect_test::expect_file![concat!(env!("CARGO_MANIFEST_DIR"), "/snapshots/fmt__debugdiagnostic.snap")].assert_eq(&render(format!("{report:?}")));
+/// #
+/// # stringify!(
+/// println!("{report:?}");
+/// # );
+/// ```
+///
+/// <pre>
+#[doc = include_str!(concat!(env!("CARGO_MANIFEST_DIR"), "/snapshots/fmt__debugdiagnostic.snap"))]
+/// </pre>
+///
+/// # Implementation Notes
+///
+/// This is currently very limited as items attached via [`.attach()`]
+/// only provide themselves as [`Demand`], but do not propagate [`.provide()`] to the attachment
+/// itself.
+/// This is due to limitations of Rust, as specialization is currently very unstable and partially
+/// unsound.
+///
+/// > For the current status regarding specialization refer to the tracking issue for [RFC 1210]
+/// > [#31844]
+///
+/// [`Provider`]: core::any::Provider
+/// [`.provide()`]: core::any::Provider::provide
+/// [`.attach()`]: crate::Report::attach
+/// [`Demand`]: core::any::Demand
+/// [`Debug`]: core::fmt::Debug
+/// [RFC 1210]: https://github.com/rust-lang/rfcs/pull/1210
+/// [#31844]: https://github.com/rust-lang/rust/issues/31844
+// TODO: remove experimental flag once specialisation is stabilized or sound or `.attach_provider()`
+//  is introduced.
+pub struct DebugDiagnostic {
```

Comment:
Yes, this is mainly here because of earlier experimentation. I like the idea, but it fell through/wasn't living up to the potential after finding out that `attach()` doesn't implement `Provider` for the type itself but instead creates a demand for the attachment.

But I think it's okay to let it stay in the lib under experimental to see if people like it.